source/common/http/conn_pool_base.cc

```diff
@@ -56,29 +56,32 @@ HttpConnPoolImplBase::HttpConnPoolImplBase(
 ConnectionPool::Cancellable*
 HttpConnPoolImplBase::newStream(Http::ResponseDecoder& response_decoder,
                                 Http::ConnectionPool::Callbacks& callbacks) {
-  AttachContext context = std::make_pair(&response_decoder, &callbacks);
-  return Envoy::ConnectionPool::ConnPoolImplBase::newStream(reinterpret_cast<void*>(&context));
+  HttpAttachContext context({&response_decoder, &callbacks});
+  return Envoy::ConnectionPool::ConnPoolImplBase::newStream(context);
 }
 
 bool HttpConnPoolImplBase::hasActiveConnections() const {
   return (!pending_requests_.empty() || (num_active_requests_ > 0));
 }
 
-ConnectionPool::Cancellable* HttpConnPoolImplBase::newPendingRequest(void* context) {
-  Http::ResponseDecoder& decoder = *reinterpret_cast<AttachContext*>(context)->first;
-  Http::ConnectionPool::Callbacks& callbacks = *reinterpret_cast<AttachContext*>(context)->second;
+ConnectionPool::Cancellable*
+HttpConnPoolImplBase::newPendingRequest(const Envoy::ConnectionPool::AttachContext& context) {
+  Http::ResponseDecoder& decoder = *static_cast<const HttpAttachContext*>(&context)->decoder_;
+  Http::ConnectionPool::Callbacks& callbacks =
+      *static_cast<const HttpAttachContext*>(&context)->callbacks_;
```

Comment:
Technically I think you should do dynamic_cast here and elsewhere. You might consider a helper that does it and returns the real object, or just a template method on the interface like we for for thread local storage slots.

Reply:
Heh, Greg had me go from reinterpret to static.  I think static cast is consistent with the style guide*.  Maybe I could add ASSERTS for dynamic cast like Greg suggested?

*https://google.github.io/styleguide/cppguide.html#Casting

Reply:
Sure dynamic_cast in ASSERT SGTM, I would still put it in a templated helper though on the interface. Can you potentially take an action to change the ThreadLocal interface to do the same thing? There might be a nice perf boost there.

Reply:
template function added.  lmk if it's not what you were looking for.
Sadly when I tried to add the dynamic cast assert it failed because not polymorphic - no virtuals in the base class). I could add some or we could leave as-is.

Reply:
Also yeah, happy to do the follow-up with TLS.

Reply:
Yes this LGTM, sorry why does the dynamic_cast fail? The virtual destructor doesn't count?